### DISCUSSION TITLE
[Feature] Official gpu / cuda support for `immich_machine_learning`

### DISCUSSION THREAD

**Discussion by simone-viozzi** [immich-app/immich#2368] [Feature Request]
### The feature

As for the current version, all the machine learning requests are handled by the CPU, with no option to enable the GPU.

Since I have a 1050ti on my server, I would like to use it to handle the models.

I actually managed to make it work by modifying the `machine-learning` dockerfile and code:

On the `Dockerfile`:

```diff
< RUN /opt/venv/bin/pip install --pre torch  -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html
---
> RUN /opt/venv/bin/pip install --pre torch  -f https://download.pytorch.org/whl/cu118
```

On `main.py`:

```diff
<             _model_cache[key] = pipeline(model=model, task=task)
---
>             _model_cache[key] = pipeline(model=model, task=task, device=0)
```

And of course you need to add the GPU capabilities to the `docker-compose`:

```yaml
    deploy:
      resources:
        reservations:
          devices:
            - driver: nvidia
              count: 1
              capabilities: [gpu]
```

What I would like is, an official and supported way to enable the GPU support, like with just an ENV var.

Thank you for this wonderful app!

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by alextran1502** [comment]
Amazing! Thank you, we will look into adding this to the next release

**Reply by HybridSarcasm** [reply]
Hi Alex. Does this have a milestone assigned yet? I have a GTX1650 waiting for something to do.

**Comment by simone-viozzi** [comment]
To add more context, with both models on a 1050Ti, it only took 3h to process 30k images I uploaded from the immich-cli.

**Reply by mirisbowring** [reply]
would this also support iGPU (QuickSync)?

**Reply by simone-viozzi** [reply]
nop, `transformers` and `sentence-transformer` need a CUDA device, so only nvidia gpus.
To run on other types of accelerators, you would need to convert the model in some ways.

**Reply by simone-viozzi** [reply]
For example, the new Face Recognition module use `onnxruntime`, that is generally more optimized to run on CPU. So the gain on running it on a GPU is a lot less than `sentence-transformer`.

**Reply by IronBeardKnight** [reply]
This is a bummer but what about the video transcoding settings inside the none alpine release of the docker container does this mean they are now useless settings?

**Comment by DrSpaldo** [comment]
I'm also very interested in this one, I am a new user adding my collection and even with turning one option on at a time (currently 'encode clip') it is smashing my server. See below

![2023-05-17 screenshot](https://github.com/immich-app/immich/assets/8526509/024d6299-0ec0-48c0-ad81-52a5fc136402)

**Reply by simone-viozzi** [reply]
Anyway, you can limit how many CPUs `immich-machine-learning` is allowed to use by adding:

```yaml
    deploy:
      resources:
        limits:
          cpus: '4'
```

To the docker-compose ([reference](https://stackoverflow.com/a/57135933/8069500)).

**Comment by martabal** [comment]
Hey, we integrated cuda acceleration in the AIO image, but I couldn't get facial recognition to work with cuda, it always uses 'CPUProvider'. Did you manage to make it work ?

**Reply by simone-viozzi** [reply]
Yes, but is kind of a mess. You can find the dockerfile and code at: https://github.com/simone-viozzi/my-server/tree/436ff2fe6cce41d02131c2dc8d8a9c6ef0c83ee7/immich. Use it as a starting point, it is not stable at all.

To make onnxruntime work on the GPU, you need to start from an image with CUDA + cublas.

Also, on my 1050ti the facial recognition model was not that fast, ~2x compared to CPU.

**Comment by martabal** [comment]
With the amazing work of @simone-viozzi , I have a working version [here](https://github.com/martabal/immich/tree/feat/support-cuda-acceleration/machine-learning), should I open a PR ?

**Reply by alextran1502** [reply]
Yes please ❤️ ! Can you also include the instruction, best would be in the documentation page on how to use the CUDA build. We will probably add a Github action to build the CUDA related image as well

**Reply by martabal** [reply]
Yes, no probelm

**Comment by IronBeardKnight** [comment]
I can see in the Unraid docker version recent release this was removed can we know why it was removed as it seems like the most logical thing to include into a project that is running any type of machine learning?

Is it as simple as adding in the variables for nvidia and running the container or has the logic within the image been removed again. 

Would love some feedback thanks in advance guys.

**Reply by IronBeardKnight** [reply]
@alextran1502 and @martabal

**Reply by martabal** [reply]
You can see the full explanation [here](https://github.com/imagegenius/docker-immich/issues/151#issuecomment-1650082889)

TL;DR : It was a custom patch and machine learning has changed a lot since then, so maintaining it was very hard (or impossible). We prefer to support it when Immich will officially support it

**Comment by anultravioletaurora** [comment]
Did this feature go stale? I'd be interested in getting it working and can contribute if needed ☺️

It seems like while we want the Dockerfile option to specify a GPU for ML workloads, there is still work needed to add the correct core ML support within Immich itself? 

I know there was another PR with a lot of code changes - maybe we can pin down what all need to change and only touch what's needed? I'd be happy to contribute to this since I passionately want to see it working, I just might need some help from others more familiar with the code base. I'm still new to this app, so apologies for my naivety 😅

**Reply by bo0tzz** [reply]
cc @mertalev

**Reply by mertalev** [reply]
I have a more recent branch that adds CUDA support, but its memory usage didn't seem normal so I didn't make a PR for it. Models took up space in both RAM and VRAM, and actually used more RAM than with normal CPU execution.

**Reply by anultravioletaurora** [reply]
oof - how much extra system RAM did it use? Like a noticeable amount? I imagine enough to hold the models in memory?

I guess I'm wondering what the trade off is be for the execution times and functionality? Obvs it can't be too high but do you think it's worth it for the results and functionality?

**Reply by mertalev** [reply]
IIRC it was using around 4gb of RAM and under 3gb of VRAM. My hunch is that it's loading the models to RAM and then to VRAM, but the models in RAM don't get deallocated. The difference in performance for me was maybe 30-50%, although this will depend on a lot of factors. 

The reason it isn't a bigger difference is that we don't do any batching for images. This doesn't hurt too much for CPUs, but for GPUs it reduces parallelism and causes a lot of overhead from CPU->GPU transfers. I'd be fine with adding CUDA support before adding batching, though, just if the memory usage was more reasonable.

**Reply by anultravioletaurora** [reply]
Ahhh I see - so adding batching for ML workloads would reduce that overhead and make it "feel" hardware accelerated?

Would it be necessary to have those models be removed from memory? Does it hinder the stability of the app at all? If it doesn't, maybe that can be a lower priority effort for this and we can save that for an update after a GPU Acceleration Support "MVP"

**Reply by mertalev** [reply]
I can see that point of view as well, but IMO this is a bug that should be sorted out first since this behavior is unexpected from a user perspective. It could affect the stability of the app if someone enables it without realizing they need to have much more RAM than they expected. If anything, a valid assumption would be that they'd need less RAM since the models would be in the GPU.

**Reply by anultravioletaurora** [reply]
Yeah that totally makes sense

So from what I'm understanding is that this effort has a couple moving parts:

- Adding CUDA support to Immich ML, using your work as a starting point (still need to tame RAM usage)
- Adding batching support for ML workloads for reduced CPU overhead
- Adding Dockerfile components to support CUDA

That's everything we'd need to make this feature shippable?

**Reply by jrasm91** [reply]
Right, but batching support for ML is complicated, since it involves updating the main codebase which does everything on a per job basis. It is unclear what the added complexity will be, but it is likely to be high, and it may or may not be worth pursuing right now. There are definitely some improvements in job throughput in general if we're able to minimize loading/unloading assets to/from memory, but it may just be too complicated to implement any time soon.

**Reply by mertalev** [reply]
I did work on a bespoke dynamic batching feature that keeps batching logic in the ML service. It basically aggregates separate requests into a batch, processes it and separates it back. I've been meaning to revisit that, let me dig it up.

**Reply by jrasm91** [reply]
Oh yeah, that would possibly work.

**Reply by owerrride** [reply]
Please tell me, is there any way to use the GPU for machine learning now?

**Reply by mertalev** [reply]
Not at the moment. My focus is currently on transitioning away from Typesense, so CUDA support will not be added until after this change is ready.

**Comment by mandusm** [comment]
Any update on this?

**Reply by mertalev** [reply]
It will most likely be in the 1.94.0 release; see [this](https://github.com/immich-app/immich/pull/5619) PR.
